xdsl/tools/tblgen_to_py.py

```diff
@@ -0,0 +1,566 @@
+            case "AnyComplex":
+                return "BaseAttr(ComplexType)"
+            case _:
+                if "AnyTypeOf" in rec.superclasses:
+                    return textwrap.dedent(f"""
+                    AnyOf(
+                        (
+                            {",".join(self._resolve_type_constraint(x["def"]) for x in rec["allowedTypes"])}
+                        )
+                    )
+                    """)
+
+                if "AllOfType" in rec.superclasses:
+                    return textwrap.dedent(f"""
+                    AllOf(
+                        (
+                            {",".join(self._resolve_type_constraint(x["def"]) for x in rec["allowedTypes"])}
+                        )
+                    )
+                    """)
+
+                if "AnyI" in rec.superclasses:
+                    return textwrap.dedent(f"""
+                    ParamAttrConstraint(
+                        IntegerType,
+                        (EqAttrConstraint(IntAttr({rec["bitwidth"]})), AnyAttr()),
+                    )
+                    """)
+
+                if "I" in rec.superclasses:
+                    return f"EqAttrConstraint(IntegerType({rec['bitwidth']}))"
+                if "SI" in rec.superclasses:
+                    return f"EqAttrConstraint(IntegerType({rec['bitwidth']}, Signedness.SIGNED))"
+                if "UI" in rec.superclasses:
+                    return f"EqAttrConstraint(IntegerType({rec['bitwidth']}, Signedness.UNSIGNED))"
+                if "Complex" in rec.superclasses:
+                    return textwrap.dedent(f"""
+                    ParamAttrConstraint(
+                        ComplexType,
+                        ({self._resolve_type_constraint(rec["elementType"]["def"])},),
+                    )
+                    """)
+
+                return "AnyAttr()"
+
+    def _resolve_prop_constraint(self, rec: TblgenRecord | str) -> str:
+        if isinstance(rec, str):
+            rec = self._get_record(rec)
+
+        if rec.name in self.attributes:
+            return f"BaseAttr({rec.name})"
+
+        match rec.name:
+            case "BoolAttr":
+                return "BaseAttr(BoolAttr)"
+            case "IndexAttr":
+                return textwrap.dedent("""
+                ParamAttrConstraint(
+                    IntegerAttr, (AnyAttr(), EqAttrConstraint(IndexType()))
+                )
+                """)
+
+            case "APIntAttr":
+                return textwrap.dedent("""
+                ParamAttrConstraint(
+                    IntegerAttr, (AnyAttr(), AnyAttr())
+                )
+                """)  # TODO can't represent APInt properly
+
+            case "StrAttr":
+                return "BaseAttr(StringAttr)"
+            case "SymbolNameAttr":
+                return "BaseAttr(SymbolNameAttr)"
+            case "UnitAttr":
+                return "EqAttrConstraint(UnitAttr())"
+            case _:
```

Comment:
Nit: Since anyway all other cases are doing a `return`, you can de indent this one

Reply:
I think I mainly kept it like this because pyright wasn't happy if the case was omitted entirely, but happy to dedent

Reply:
If pyright wasn't happy let's not make it angry then ;)